edgelet/edgelet-kube/src/module/authentication.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use std::cell::RefCell;
+use std::sync::{Arc, Mutex};
+
+use futures::future::{Either, IntoFuture};
+use futures::{future, Future, Stream};
+use hyper::service::Service;
+use hyper::{Body, Request};
+use log::Level;
+use typed_headers::{Authorization, HeaderMapExt};
+
+use edgelet_core::AuthId;
+use edgelet_utils::log_failure;
+use kube_client::{Client as KubeClient, Error as KubeClientError, TokenSource};
+
+use crate::constants::EDGE_ORIGINAL_MODULEID;
+use crate::error::Error;
+use crate::{KubeModuleRuntime, KubeRuntimeData};
+
+pub fn authenticate<T, S>(
+    runtime: &KubeModuleRuntime<T, S>,
+    req: &Request<Body>,
+) -> impl Future<Item = AuthId, Error = Error>
+where
+    T: TokenSource + 'static,
+    S: Service + Send + 'static,
+    S::ReqBody: From<Vec<u8>>,
+    S::ResBody: Stream,
+    Body: From<S::ResBody>,
+    S::Error: Into<KubeClientError>,
+    S::Future: Send,
+{
+    req.headers()
+        .typed_get::<Authorization>()
+        .map(|auth| {
+            auth.and_then(|auth| {
+                auth.as_bearer().map(|token| {
+                    let client_copy = runtime.client();
+                    let namespace = runtime.namespace().to_owned();
+                    let fut = runtime
+                        .client()
+                        .lock()
+                        .expect("Unexpected lock error")
+                        .borrow_mut()
+                        .token_review(runtime.namespace(), token.as_str())
+                        .map_err(|err| {
+                            log_failure(Level::Warn, &err);
+                            Error::from(err)
+                        })
+                        .and_then(move |token_review| {
+                            token_review
+                                .status
+                                .as_ref()
+                                .filter(|status| status.authenticated.filter(|x| *x).is_some())
+                                .and_then(|status| {
+                                    status.user.as_ref().and_then(|user| user.username.clone())
+                                })
+                                .map_or(Either::A(future::ok(AuthId::None)), |name| {
+                                    Either::B(get_module_original_name(
+                                        &client_copy,
+                                        &namespace,
+                                        &name,
+                                    ))
+                                })
+                        });
+
+                    Either::A(fut)
+                })
+            })
+            .unwrap_or_else(|| Either::B(future::ok(AuthId::None)))
+        })
+        .map_err(Error::from)
+        .into_future()
+        .flatten()
+}
+
+fn get_module_original_name<T, S>(
+    client: &Arc<Mutex<RefCell<KubeClient<T, S>>>>,
```

Comment:
We try to avoid RefCell where possible because the borrow rules are then checked at runtime. Are you sure it is absolutely needed here?

Reply:
As we discussed offline `RefCell` is how `KubeModuleRuntime` stores `KubeClient` instance in order to pass `&mut self` inside hyper `Service::call()`.